nexus/types/src/deployment/tri_map.rs

```diff
@@ -0,0 +1,509 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{
+    borrow::Borrow,
+    collections::{hash_map, BTreeSet, HashMap},
+    fmt,
+    hash::Hash,
+};
+
+use derive_where::derive_where;
+use serde::{Deserialize, Serialize, Serializer};
+
+/// An append-only 1:1:1 (trijective) map for three keys and a value.
```

Comment:
This is append-only for now -- we may add the ability to remove entries in the future if really necessary. (I think at least at the moment it isn't necessary either here or in `BlueprintBuilder`.)